405 Method Not Allowed should include an Allow header

kellymclaughlin opened the pull request:
A 405 Method Not Allowed response should include an Allow header: https://tools.ietf.org/html/rfc7231#section-6.5.5

kellymclaughlin committed:
Include Allow header with 405 responses

Include an Allow header listing the resource's supported methods with
405 responses per section 6.5.5 of RFC 7231.

patrickt commented:
+1 875b9c2

reiddraper commented on src/Airship/Internal/Decision.hs:
```diff
@@ -192,7 +192,9 @@ b10 r@Resource{..} = do
         then b09 r
-        else lift $ halt HTTP.status405
+        else do
+            lift $ addResponseHeader ("Allow", renderHeader allowed)
```
`renderHeader` is for HTTP media types, right? What does this end up looking like?

reiddraper replied:
`renderHeader :: MediaType -> ByteString`

kellymclaughlin committed:
Use intercalate to compose Allow header

Since Method in Network.HTTP.Types is type alias for ByteString use
Data.ByteString.intercalate to compose the Allow header instead of
renderHeader which is intended for media types.

kellymclaughlin commented:
Here's some output of a curl POST to the default airship root resource to show how the header appears using this branch:

```
 $ curl localhost:3000 -v --data "abc"
* Rebuilt URL to: localhost:3000/
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 3000 (#0)
> POST / HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.43.0
> Accept: */*
> Content-Length: 3
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 3 out of 3 bytes
< HTTP/1.1 405 Method Not Allowed
< Transfer-Encoding: chunked
< Date: Thu, 16 Jul 2015 23:42:43 GMT
< Server: Warp/3.0.13.1
< Allow: GET,HEAD
< Airship-Trace: b13,b12,b11,b10
< Airship-Quip: WARNING: ulimit -n is 1024
< 
* Connection #0 to host localhost left intact
```

dkubb commented:
@kellymclaughlin that looks great, exactly what I would expect.
